### PR TITLE
pairs2hashMapArray

### PR DESCRIPTION
**Pull request by shkel**

### CODE REVIEW

File: lib/ds/hashmap.flow [hunk 1]
```diff
@@ -221,6 +222,16 @@ pairs2hashMapWithLoadFactor(pairs : [Pair<?, ??>], hash : (?) -> int, loadFactor
 	hashMap
 }
 
+pairs2hashMapArray(pairs : [Pair<?, [??]>], hash : (?) -> int) -> HashMap<?, [??]> {
+	len = cast(cast(length(pairs) : int -> double) / defaultHashMapLoadFactor : double -> int) * 2;
```
Comment (alstrup) [inline lib/ds/hashmap.flow]: i2d instead of cast

Comment (alstrup) [reply lib/ds/hashmap.flow]: and floor the other way

File: lib/ds/hashmap.flow [hunk 2]
```diff
@@ -221,6 +222,16 @@ pairs2hashMapWithLoadFactor(pairs : [Pair<?, ??>], hash : (?) -> int, loadFactor
...
+	hashMap = makeHashMapOfVolume(hash, len, defaultHashMapLoadFactor);
+	iter(pairs, \pair -> {
```
Comment (alstrup) [inline lib/ds/hashmap.flow]: Why is this not a fold? Isn't the hashmap functional?

Comment (shkel) [reply lib/ds/hashmap.flow]: copy-paste from pairs2hashMap()

File: lib/ds/hashmap.flow [hunk 3]
```diff
@@ -221,6 +223,18 @@ pairs2hashMapWithLoadFactor(pairs : [Pair<?, ??>], hash : (?) -> int, loadFactor
...
 }
 
+pairs2hashMapArray(pairs : [Pair<?, [??]>], hash : (?) -> int) -> HashMap<?, [??]> {
+	len = floor(i2d(length(pairs)) / defaultHashMapLoadFactor) * 2;
+	hashMap = makeHashMapOfVolume(hash, len, defaultHashMapLoadFactor);
+	fold(pairs, hashMap, \newHashMap, pair -> { 
+		setHashMap(
+			newHashMap,
+			pair.first,
+			eitherMap(lookupHashMap(hashMap, pair.first), \arr -> concat(arr, pair.second), pair.second)
```
Comment (alstrup) [inline lib/ds/hashmap.flow]: lookupHashMapDef  seems like your friend: Just return the empty array as def, and then you can concat?

Comment (shkel) [reply lib/ds/hashmap.flow]: lookupHashMapDef + concat produces less garbage?

Comment (alstrup) [reply lib/ds/hashmap.flow]: Probably similar garbage but easier to read

**Pull request merged**